In [3]:
pip install sklearn

  Using cached sklearn-0.0.post12.tar.gz (2.6 kB)
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'error'
Note: you may need to restart the kernel to use updated packages.


  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [15 lines of output]
      The 'sklearn' PyPI package is deprecated, use 'scikit-learn'
      rather than 'sklearn' for pip commands.
      
      Here is how to fix this error in the main use cases:
      - use 'pip install scikit-learn' rather than 'pip install sklearn'
      - replace 'sklearn' by 'scikit-learn' in your pip requirements files
        (requirements.txt, setup.py, setup.cfg, Pipfile, etc ...)
      - if the 'sklearn' package is used by one of your dependencies,
        it would be great if you take some time to track which package uses
        'sklearn' instead of 'scikit-learn' and report it to their issue tracker
      - as a last resort, set the environment variable
        SKLEARN_ALLOW_DEPRECATED_SKLEARN_PACKAGE_INSTALL=True to avoid this error
      
      More information is available at
      https://github.com/scikit-learn/sklearn-pypi-packag

In [4]:
from sklearn.datasets import fetch_openml
import pandas as pd

data = fetch_openml(name = 'credit-g', version = 1, as_frame = True)
df= data.frame
df.head()

checking_status  duration  ... foreign_worker class
0              <0         6  ...            yes  good
1        0<=X<200        48  ...            yes   bad
2     no checking        12  ...            yes  good
3              <0        42  ...            yes  good
4              <0        24  ...            yes   bad

[5 rows x 21 columns]

In [5]:
df.columns

Index(['checking_status', 'duration', 'credit_history', 'purpose',
       'credit_amount', 'savings_status', 'employment',
       'installment_commitment', 'personal_status', 'other_parties',
       'residence_since', 'property_magnitude', 'age', 'other_payment_plans',
       'housing', 'existing_credits', 'job', 'num_dependents', 'own_telephone',
       'foreign_worker', 'class'],
      dtype='object')

# German Credit Dataset – Feature Metadata

| Column                  | Description |
|-------------------------|-------------|
| `checking_status`       | Status of the existing checking account (e.g., `<0 DM`, `1-200 DM`, `>200 DM`) |
| `duration`              | Duration in months of the credit (loan term) |
| `credit_history`        | Record of the applicant’s past credit performance |
| `purpose`               | Purpose of the loan (e.g., car, education, furniture, radio/TV) |
| `credit_amount`         | The loan amount requested (in Deutsche Marks) |
| `savings_status`        | Status of savings account/bonds |
| `employment`            | Duration of current employment (e.g., `1-4 years`, `>7 years`) |
| `installment_commitment`| Installment rate as a percentage of disposable income |
| `personal_status`       | Gender and marital status (e.g., male-single, female-divorced) |
| `other_parties`         | Other parties involved (e.g., guarantor, co-applicant) |
| `residence_since`       | Number of years at current residence |
| `property_magnitude`    | Type of owned property (e.g., real estate, life insurance) |
| `age`                   | Age of the applicant in years |
| `other_payment_plans`   | Other existing payment plans (e.g., bank, stores) |
| `housing`               | Housing type (e.g., own, rent, for free) |
| `existing_credits`      | Number of existing credits at this bank |
| `job`                   | Nature of job or employment class |
| `num_dependents`        | Number of people being financially supported |
| `own_telephone`         | Whether the applicant owns a phone (yes/no) |
| `foreign_worker`        | Whether the applicant is a foreign worker (yes/no) |
| `class`                 | Target variable – **Credit risk** (`good` or `bad`) |


### Understanding the Data

In [6]:
df.nunique()

checking_status             4
duration                   33
credit_history              5
purpose                    10
credit_amount             921
savings_status              5
employment                  5
installment_commitment      4
personal_status             4
other_parties               3
residence_since             4
property_magnitude          4
age                        53
other_payment_plans         3
housing                     3
existing_credits            4
job                         4
num_dependents              2
own_telephone               2
foreign_worker              2
class                       2
dtype: int64

In [7]:
for col in df.columns:
    print(col,':', df[col].unique())

checking_status : ['<0', '0<=X<200', 'no checking', '>=200']
Categories (4, object): ['0<=X<200', '<0', '>=200', 'no checking']
duration : [ 6 48 12 42 24 36 30 15  9 10  7 60 18 45 11 27  8 54 20 14 33 21 16  4
 47 13 22 39 28  5 26 72 40]
credit_history : ['critical/other existing credit', 'existing paid', 'delayed previously', 'no credits/all paid', 'all paid']
Categories (5, object): ['all paid', 'critical/other existing credit', 'delayed previously', 'existing paid', 'no credits/all paid']
purpose : ['radio/tv', 'education', 'furniture/equipment', 'new car', 'used car', 'business', 'domestic appliance', 'repairs', 'other', 'retraining']
Categories (10, object): ['domestic appliance', 'new car', 'used car', 'business', ..., 'other', 'radio/tv', 'repairs', 'retraining']
credit_amount : [ 1169  5951  2096  7882  4870  9055  2835  6948  3059  5234  1295  4308
  1567  1199  1403  1282  2424  8072 12579  3430  2134  2647  2241  1804
  2069  1374   426   409  2415  6836  1913  4020  5866

In [8]:
df.isnull().sum()

checking_status           0
duration                  0
credit_history            0
purpose                   0
credit_amount             0
savings_status            0
employment                0
installment_commitment    0
personal_status           0
other_parties             0
residence_since           0
property_magnitude        0
age                       0
other_payment_plans       0
housing                   0
existing_credits          0
job                       0
num_dependents            0
own_telephone             0
foreign_worker            0
class                     0
dtype: int64

In [9]:
df.isna().sum()

checking_status           0
duration                  0
credit_history            0
purpose                   0
credit_amount             0
savings_status            0
employment                0
installment_commitment    0
personal_status           0
other_parties             0
residence_since           0
property_magnitude        0
age                       0
other_payment_plans       0
housing                   0
existing_credits          0
job                       0
num_dependents            0
own_telephone             0
foreign_worker            0
class                     0
dtype: int64

In [10]:
df[df['class'].isna()][:10]

Empty DataFrame
Columns: [checking_status, duration, credit_history, purpose, credit_amount, savings_status, employment, installment_commitment, personal_status, other_parties, residence_since, property_magnitude, age, other_payment_plans, housing, existing_credits, job, num_dependents, own_telephone, foreign_worker, class]
Index: []

In [11]:
df.head()

checking_status  duration  ... foreign_worker class
0              <0         6  ...            yes  good
1        0<=X<200        48  ...            yes   bad
2     no checking        12  ...            yes  good
3              <0        42  ...            yes  good
4              <0        24  ...            yes   bad

[5 rows x 21 columns]

In [12]:
df.dtypes

checking_status           category
duration                     int64
credit_history            category
purpose                   category
credit_amount                int64
savings_status            category
employment                category
installment_commitment       int64
personal_status           category
other_parties             category
residence_since              int64
property_magnitude        category
age                          int64
other_payment_plans       category
housing                   category
existing_credits             int64
job                       category
num_dependents               int64
own_telephone             category
foreign_worker            category
class                     category
dtype: object

2. PreProcessing and Pipelining

In [13]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix

In [14]:
df['class'] = df['class'].map({'good': 0, 'bad': 1})

X = df.drop('class', axis = 1)
y = df['class']

category_cols = X.select_dtypes(include = 'category').columns.tolist()
numeric_cols = X.select_dtypes(exclude = 'category').columns.tolist()

In [15]:
print(numeric_cols)

['duration', 'credit_amount', 'installment_commitment', 'residence_since', 'age', 'existing_credits', 'num_dependents']


In [16]:
preprocessor = ColumnTransformer([
    ('cat', OneHotEncoder(handle_unknown='ignore'), category_cols),
    ('num', StandardScaler(), numeric_cols)
])

pipeline = Pipeline([
    ('prep',preprocessor),
    ('clf', LogisticRegression(max_iter = 1000))
])

##### 3. Train-test split and model fit

In [17]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42, stratify = y)
pipeline.fit(X_train, y_train)

Pipeline(steps=[('prep',
                 ColumnTransformer(transformers=[('cat',
                                                  OneHotEncoder(handle_unknown='ignore'),
                                                  ['checking_status',
                                                   'credit_history', 'purpose',
                                                   'savings_status',
                                                   'employment',
                                                   'personal_status',
                                                   'other_parties',
                                                   'property_magnitude',
                                                   'other_payment_plans',
                                                   'housing', 'job',
                                                   'own_telephone',
                                                   'foreign_worker']),
                                                 ('num', StandardScaler(),
                                                  ['duration', 'credit_amount',
                                                   'installment_commitment',
                                                   'residence_since', 'age',
                                                   'existing_credits',
                                                   'num_dependents'])])),
                ('clf', LogisticRegression(max_iter=1000))])

##### 5. Evaluation

In [18]:
y_pred = pipeline.predict(X_test)

print('classification_report:')
print(classification_report(y_test, y_pred))

print('Confusion matrix:')
print(confusion_matrix(y_test, y_pred))

classification_report:
              precision    recall  f1-score   support

           0       0.82      0.89      0.85       140
           1       0.67      0.53      0.59        60

    accuracy                           0.78       200
   macro avg       0.74      0.71      0.72       200
weighted avg       0.77      0.78      0.77       200

Confusion matrix:
[[124  16]
 [ 28  32]]


#### Insights:

1. The dataset is imbalanced (140 good : 60 bad), creating a 7:3 split that biases the model toward predicting good credit.
2. As expected, the model performs well on the majority class (good credit), achieving high precision and recall.
3. However, recall for the minority class (bad credit) is only 53%, meaning nearly half of potential defaulters go undetected — a serious risk for lenders.

#### Recommendation:

- Apply resampling techniques to mitigate class imbalance:
  - **SMOTE** to synthetically generate more bad credit samples
  - **Undersampling** the good credit class for a balanced training set
- Use **`class_weight='balanced'`** in your model to penalize misclassifying defaulters more heavily
- Consider **threshold tuning** to further improve recall for class 1


In [19]:
df['class'].value_counts(normalize = True)

class
0    0.7
1    0.3
Name: proportion, dtype: float64

### Undersampling to Balance the Dataset

In [20]:
pip install imblearn

Note: you may need to restart the kernel to use updated packages.


In [21]:
from sklearn.model_selection import train_test_split
from imblearn.under_sampling import RandomUnderSampler
from collections import Counter

In [22]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, stratify = y, random_state=42
)

print("Before undersampling:", Counter(y_train))

undersampler = RandomUnderSampler(random_state=42)
X_resampled, y_resampled= undersampler.fit_resample(X_train, y_train)

print("After undersampling:", Counter(y_resampled))

Before undersampling: Counter({0: 560, 1: 240})
After undersampling: Counter({0: 240, 1: 240})


In [23]:
pipeline.fit(X_resampled, y_resampled)

Pipeline(steps=[('prep',
                 ColumnTransformer(transformers=[('cat',
                                                  OneHotEncoder(handle_unknown='ignore'),
                                                  ['checking_status',
                                                   'credit_history', 'purpose',
                                                   'savings_status',
                                                   'employment',
                                                   'personal_status',
                                                   'other_parties',
                                                   'property_magnitude',
                                                   'other_payment_plans',
                                                   'housing', 'job',
                                                   'own_telephone',
                                                   'foreign_worker']),
                                                 ('num', StandardScaler(),
                                                  ['duration', 'credit_amount',
                                                   'installment_commitment',
                                                   'residence_since', 'age',
                                                   'existing_credits',
                                                   'num_dependents'])])),
                ('clf', LogisticRegression(max_iter=1000))])

In [24]:
y_pred = pipeline.predict(X_test)

print('classification_report:')
print(classification_report(y_test, y_pred))

print('Confusion matrix:')
print(confusion_matrix(y_test, y_pred))

classification_report:
              precision    recall  f1-score   support

           0       0.88      0.68      0.77       140
           1       0.51      0.78      0.62        60

    accuracy                           0.71       200
   macro avg       0.70      0.73      0.69       200
weighted avg       0.77      0.71      0.72       200

Confusion matrix:
[[95 45]
 [13 47]]


1. The overall performance of the model dropped

Recommendation: 
Try oversampling
